ens/_normalization.py

```diff
@@ -0,0 +1,512 @@
+from enum import (
+    Enum,
+)
+import json
+import os
+from sys import (
+    version_info,
+)
+from typing import (
+    Any,
+    Dict,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+from pyunormalize import (
+    NFC,
+    NFD,
+)
+
+from .exceptions import (
+    InvalidName,
+)
+
+# TODO: remove once web3 supports python>=3.8
+if version_info >= (3, 8):
+    from typing import (
+        Literal,
+    )
+else:
+    from typing_extensions import (  # type: ignore
+        Literal,
+    )
+
+
+def _json_list_mapping_to_dict(
+    f: Dict[str, Any],
+    list_mapped_key: str,
+) -> Dict[str, Any]:
+    """
+    Takes a `[key, [value]]` mapping from the original ENS spec json files and turns it
+    into a `{key: value}` mapping.
+    """
+    f[list_mapped_key] = {k: v for k, v in f[list_mapped_key]}
+    return f
+
+
+# get the normalization spec json files downloaded from links in ENSIP-15
+# https://docs.ens.domains/ens-improvement-proposals/ensip-15-normalization-standard
+specs_dir_path = os.path.abspath(os.path.join(os.path.dirname(__file__), "specs"))
+with open(os.path.join(specs_dir_path, "normalization_spec.json")) as spec:
+    f = json.load(spec)
+
+    NORMALIZATION_SPEC = _json_list_mapping_to_dict(f, "mapped")
+    # clean `FE0F` (65039) from entries since it's optional
+    for e in NORMALIZATION_SPEC["emoji"]:
+        if 65039 in e:
+            for i in range(e.count(65039)):
+                e.remove(65039)
+
+with open(os.path.join(specs_dir_path, "nf.json")) as nf:
+    f = json.load(nf)
+    NF = _json_list_mapping_to_dict(f, "decomp")
+
+
+# --- Classes -- #
+
+
+class TokenType(Enum):
+    EMOJI = "emoji"
+    TEXT = "text"
+
+
+class Token:
+    type: Literal[TokenType.TEXT, TokenType.EMOJI]
+    _original_text: str
+    _original_codepoints: List[int]
+    _normalized_codepoints: Optional[List[int]] = None
+
+    restricted: bool = False
+
+    def __init__(self, original_text: str, original_codepoints: List[int]) -> None:
+        self._raw_text = original_text
+        self._original_codepoints = original_codepoints
+
+    @property
+    def codepoints(self) -> List[int]:
+        return (
+            self._normalized_codepoints
+            if self._normalized_codepoints
+            else self._original_codepoints
+        )
+
+    @property
+    def text(self) -> str:
+        return _cps_to_text(self.codepoints)
```

Comment:
another nit: Since the spec uses the word 'codepoints' so often, it might be nice to make this method name `codepoints_to_text`. Feel free to take it or leave it though :)

Reply:
Noted 👌🏼